src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -441,11 +461,131 @@ internal IEnumerable<IEdmNavigationProperty> GetSelectedNavigationProperties(IEd
             return selectedNavigationProperties.Distinct();
         }
 
+        /// <summary>
+        /// Gets all selected structured properties at the current node.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The selected structured properties at this node level.</returns>
+        /// <remarks>Result does not include the dynamic properties, which can be retrieved by <code>GetSelectedDynamicProperties</code></remarks>
+        internal IEnumerable<IEdmProperty> GetSelectedProperties(IEdmStructuredType structuredType)
+        {
+            if (this.selectionType == SelectionType.Empty)
+            {
+                return EmptyEdmProperties;
+            }
+
+            // We cannot determine the selected properties without the user model. This means we won't be computing the missing properties.
+            if (structuredType == null)
+            {
+                return EmptyEdmProperties;
+            }
+
+            if (this.selectionType == SelectionType.EntireSubtree || this.hasWildcard)
+            {
+                return structuredType.Properties();
+            }
+
+            Debug.Assert(this.selectedProperties != null, "selectedProperties != null");
+
+            // Get properties selected, and filter out unrecognized properties.
+            return this.selectedProperties.Select(structuredType.FindProperty).OfType<IEdmProperty>();
+        }
+
+        /// <summary>
+        /// Return names of dynamic properties.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The names of dynamic properties.</returns>
+        internal IEnumerable<string> GetSelectedDynamicProperties(IEdmStructuredType structuredType)
+        {
+            IEnumerable<string> dynamicProperties = Enumerable.Empty<string>();
+
+            if (this.selectedProperties != null && this.selectedProperties.Count > 0)
+            {
+                dynamicProperties = this.selectedProperties.Select(
+                    prop => structuredType.FindProperty(prop) == null
+                    ? prop
+                    : null)
+                    .OfType<string>();
+            }
+
+            return dynamicProperties;
+        }

```

Comment:
These two methods should be deleted if they are no longer used.

Reply:
cleaned up GetSelectedDynamicProperties.
GetSelectedProperties is still used by the test w/o using resource state.

---
In reply to: [233187531](https://github.com/OData/odata.net/pull/1106#discussion_r233187531) [](ancestors = 233187531)

Reply:
I don't like having methods (even internal) that are only used by tests. Is there another way the test can get the selected properties it is validating?

---
In reply to: [233297316](https://github.com/OData/odata.net/pull/1106#discussion_r233297316) [](ancestors = 233297316,233187531)